build-now.lua

```diff
@@ -0,0 +1,555 @@
+    [df.construction_type.UpDownStair] = df.tiletype.ConstructedStairUD,
+    [df.construction_type.Ramp] = df.tiletype.ConstructedRamp,
+}
+-- fill in all the track mappings, which have nice consistent naming conventions
+for i,v in ipairs(df.construction_type) do
+    if type(v) ~= 'string' then goto continue end
+    local _, _, base, dir = v:find('^(TrackR?a?m?p?)([NSEW]+)')
+    if base == 'Track' then
+        const_to_tile[i] = df.tiletype['ConstructedFloorTrack'..dir]
+    elseif base == 'TrackRamp' then
+        const_to_tile[i] = df.tiletype['ConstructedRampTrack'..dir]
+    end
+    ::continue::
+end
+
+local function set_tiletype(pos, tt)
+    local block = dfhack.maps.ensureTileBlock(pos)
+    block.tiletype[pos.x%16][pos.y%16] = tt
+    if tt == df.tiletype.ConstructedPillar then
+        block.designation[pos.x%16][pos.y%16].outside = 0
+    end
+    -- all tiles below this one are now "inside"
+    for z = pos.z-1,0,-1 do
+        block = dfhack.maps.ensureTileBlock(pos.x, pos.y, z)
+        if not block or block.designation[pos.x%16][pos.y%16].outside == 0 then
+            return
+        end
+        block.designation[pos.x%16][pos.y%16].outside = 0
+    end
+end
+
+local function adjust_tile_above(pos_above, item, construction_type)
+    if not dfhack.maps.ensureTileBlock(pos_above) then return end
+    local tt_above = dfhack.maps.getTileType(pos_above)
+    local shape_above = df.tiletype.attrs[tt_above].shape
+    if shape_above ~= df.tiletype_shape.EMPTY
+            and shape_above == df.tiletype_shape.RAMP_TOP then
+        return
+    end
+    if construction_type == df.construction_type.Wall then
+        create_and_link_construction(pos_above, item, true)
+        set_tiletype(pos_above, df.tiletype.ConstructedFloor)
+    elseif df.construction_type[construction_type]:find('Ramp') then
+        set_tiletype(pos_above, df.tiletype.RampTop)
+    end
+end
+
+-- add new construction to the world list and manage tiletype conversion
+local function build_construction(bld)
+    -- remember required metadata and get rid of building used for designation
+    local item = bld.contained_items[0].item
+    local pos = copyall(item.pos)
+    local construction_type = bld.type
+    dfhack.buildings.deconstruct(bld)
+
+    -- add entries to df.global.world.constructions and adjust tiletypes for
+    -- the construction itself
+    create_and_link_construction(pos, item, false)
+    set_tiletype(pos, const_to_tile[construction_type])
+    if construction_type == df.construction_type.Wall then
+        dig_now.link_adjacent_smooth_walls(pos)
+    end
+
+    -- for walls and ramps with empty space above, adjust the tile above
+    if construction_type == df.construction_type.Wall
+            or df.construction_type[construction_type]:find('Ramp') then
+        adjust_tile_above(xyz2pos(pos.x, pos.y, pos.z+1), item,
+                          construction_type)
+    end
```

Comment:
I found a weird bug that I think occurs in the section I've highlighted:

1. Run `createitem BAR IRON 3` (or have a small enough quantity of build materials that you can count easily)
2. Designate an up stair on the surface (b-C-u)
3. Run `build-now`
4. Designate a down stair right above the up stair (b-C-d)
5. Run `build-now`

This leaves behind the building item next to the staircase (its position isn't consistent, but it tends to be 1 tile away), which still has the TSK indicator until the game is advanced:

![image](https://user-images.githubusercontent.com/3719547/130010552-7e28d848-102e-4e71-8694-3c987630b333.png)

From some debugging, I confirmed that:
- `dfhack.items.remove` is returning false
- The position of the item is set to the position of the down stairs at the top of this function (where `item.pos` is retrieved for the first time)
- The position of the item has changed to be _next_ to the down stairs _before_ `remove()` is called. I'm not seeing where this happens yet.

Reply:
Ok, evidently this was easier to narrow down than I thought. The build item is moved in the call to `dfhack.buildings.deconstruct()`. This also happens when building the up stair, but that ends up putting the building item on the ground instead of the air, so `dfhack.items.remove()` works.

From a brief look at `deconstruct()`, I think DF's logic in `bld->deconstructItems(false, false)` might be related to the issue, but calling that function alone for a designated down stair doesn't reproduce the issue.

It's possible that a fix might belong in `deconstruct()` instead of build-now.

Reply:
I think I see what's going on here. `bld->deconstructItems(false, false);` will attempt to put the item on adjacent ground, and when it does, there is no problem. If there is no adjacent ground, then it places the object in mid-air and adds a `general_ref_projectile` to indicate that it is falling. This looks like completely rational behavior and not a bug. The problem is that `dfhack.items.remove()` fails for items that have a `PROJECTILE` general ref because `detachItem()` doesn't handle them.

Is there any reason why I shouldn't implement projectile handling in `detachItem()`? From inspection all it would take is deleting the link from `world->proj_list` and removing the general ref. I'll code it up and try it out.

Reply:
Thanks for tracking this down!

https://github.com/DFHack/dfhack/commit/b992b04f0bb7ffef8e7fb152967d438f398543d7 was where the projectile check was added. It doesn't go into much detail about that, but previously there were _no_ checks for unhandled general refs, so I think handling projectiles the way you did in dfhack/dfhack#1926 is reasonable.